include/contraction/contractionGraph.hpp

```diff
@@ -40,220 +40,304 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+        min_edge_id = 0;
+    }
+
+    int64_t get_next_id() {
+        return --min_edge_id;
+    }
+
+    /*!
+        @brief get the vertex descriptors of adjacent vertices of *v*
+        @param [in] v vertex_descriptor
+        @return Identifiers<V>: The set of vertex descriptors adjacent
+        to the given vertex *v*
+    */
+    Identifiers<V> find_adjacent_vertices(V v) const {
+        EO_i out, out_end;
+        EI_i in, in_end;
+        Identifiers<V> adjacent_vertices;
+
+        for (
+            boost::tie(out, out_end) = out_edges(v, this->graph);
+            out != out_end;
+            ++out
+        )
+        adjacent_vertices += this->adjacent(v, *out);
+
+        for (
+            boost::tie(in, in_end) = in_edges(v, this->graph);
+            in != in_end;
+            ++in
+        )
+        adjacent_vertices += this->adjacent(v, *in);
+
+        return adjacent_vertices;
+    }
+
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    std::vector<E> get_shortcuts() {
+        pgrouting::Identifiers<E> eids;
+        E_i e, e_end;
+        for (boost::tie(e, e_end) = edges(this->graph); e != e_end; ++e) {
+            if (this->graph[*e].id < 0) {
+                eids += *e;
+                pgassert(!(this->graph[*e]).get_contracted_vertices().empty());
+            } else {
+                pgassert((this->graph[*e]).get_contracted_vertices().empty());
+            }
+        }
+        std::vector<E> o_eids(eids.begin(), eids.end());
+        std::sort(
+            o_eids.begin(),
+            o_eids.end(),
+            [&](E lhs, E rhs) {
+                return -1*(this->graph)[lhs].id < -1*(this->graph)[rhs].id;
+            });
+        return o_eids;
+    }
+
+    /*!
+        @brief vertices with at least one contracted vertex
+        @result The vids Identifiers with at least one contracted vertex
+    */
+    Identifiers<int64_t> get_modified_vertices() {
+        Identifiers<int64_t> vids;
+        for (auto v : boost::make_iterator_range(
+            boost::vertices(this->graph))) {
+            if ((this->graph[v]).has_contracted_vertices()) {
+                vids += (this->graph[v]).get_id();
+            }
+        }
+        return vids;
+    }
+
+    /*!
+        @brief get the edge with minimum cost between two vertices
+        @param [in] u vertex_descriptor of source vertex
+        @param [in] v vertex_descriptor of target vertex
+        @return E: The edge descriptor of the edge with minimum cost
+    */
+    std::tuple<double, Identifiers<int64_t>, bool> get_min_cost_edge(V u, V v) {
+        E min_edge;
+        Identifiers<int64_t> contracted_vertices;
+        double min_cost = (std::numeric_limits<double>::max)();
+        bool found = false;
+
+        if (this->is_directed()) {
+            BGL_FORALL_OUTEDGES_T(u, e, this->graph, G) {
```

Comment:
is this macro still useful/performant? Can you consider using a for loop with iterator_range like other functions?

Reply:
Ca semble être la norme encore dans la dernière version : https://www.boost.org/doc/libs/1_78_0/boost/graph/iteration_macros.hpp

Reply:
I think the code is just moved without any modification there. I prefer to let it as it is (especially if pgrouting uses this macro elsewhere).